DQM/DTMonitorModule/src/DTDataIntegrityTask.cc

```diff
@@ -484,20 +659,341 @@ void DTDataIntegrityTask::bookHistos(DQMStore::IBooker & ibooker, string folder,
+
+// ******************uROS******************** //
+void DTDataIntegrityTask::processuROS(DTuROSROSData & data, int fed, int uRos){
+
+  neventsuROS++; // FIXME: implement a counter which makes sense
+
+      LogTrace("DTRawToDigi|DTDQM|DTMonitorModule|DTDataIntegrityTask")
+        << "[DTDataIntegrityTask]: " << neventsuROS << " events analyzed by processuROS" << endl;
+
+  if(mode == 3 || mode ==1) return; //Avoid duplication of Info in FEDIntegrity_EvF
+
+  MonitorElement* uROSSummary = nullptr;
+  uROSSummary = summaryHistos["uROSSummary"][fed];
+
+  MonitorElement* uROSStatus = nullptr;
+  uROSStatus = fedHistos["uROSStatus"][fed];
+
+  if(!uROSSummary) {
+    LogError("DTRawToDigi|DTDQM|DTMonitorModule|DTDataIntegrityTask") <<
+        "Trying to access non existing ME at FED " << fed  <<
+        std::endl;
+    return;
+  }
+
+  // Summary of all Link errors
+  MonitorElement* uROSError0 = nullptr;
+  MonitorElement* uROSError1 = nullptr;
+  MonitorElement* uROSError2 = nullptr;
+  if(mode <= 2){
+    uROSError0 = urosHistos["uROSError0"][fed][uRos];
+    uROSError1 = urosHistos["uROSError1"][fed][uRos];
+    uROSError2 = urosHistos["uROSError2"][fed][uRos];
+
+    if ((!uROSError2) || (!uROSError1) || (!uROSError0) ) {
+      LogError("DTRawToDigi|DTDQM|DTMonitorModule|DTDataIntegrityTask") <<
+        "Trying to access non existing ME at uROS " << uRos  <<
+        std::endl;
+     return;
+    }
+  }
+
+  // uROS errors
+   int sumOKs =0 ;
+   for (unsigned int link = 0; link<72; ++link){
+      for (unsigned int flag = 0; flag<5; ++flag){
+	if((data.getokxflag(link)>>flag) & 0x1) { // Undefined Flag 1-4 64bits word for each MTP (12 channels)i
+		if (flag==0) {sumOKs+=1; // Not OK flag
+				flag=5; //move it to the 5th bin
+			}
+	        if(link<24) uROSError0->Fill(flag-1,link); //bins start at 1 despite labeling
+	        else if(link<48) uROSError1->Fill(flag-1,link-23);
+        	else uROSError2->Fill(flag-1,link-47);
+	}
+      }
+   }
+   
+
+// Fill the ROSSummary (1 per FED) histo
+    if(sumOKs!=0) uROSSummary->Fill(12, uRos); //bins start at 1 despite labeling
+// Need to fill bin 13 Ev ID Mismatch?
+
+// Global Errors for uROS
+   for (unsigned int flag = 4; flag<16; ++flag){
+      if((data.getuserWord()>>flag) & 0x1 ) {
+		uROSSummary->Fill(flag-4, uRos);
+		uROSStatus->Fill(flag-4, uRos); //duplicated info?
+	}
+   }
+
+  // ROS error
+  for (unsigned int icounter = 0; icounter<data.geterrors().size(); ++icounter){
+   int link = data.geterrorROBID(icounter);
+   int tdc = data.geterrorTDCID(icounter);
+   int error =  data.geterror(icounter);
+    int tdcError_ROSSummary = 0;
+    int tdcError_ROSError = 0;
+    int tdcError_TDCHisto = 0;
+
+    if(error & 0x4000 ) {
+      LogTrace("DTRawToDigi|DTDQM|DTMonitorModule|DTDataIntegrityTask")
+        << " ROS " << uRos << " ROB " << link
+        << " Internal fatal Error 4000 in TDC " << error << endl;
+
+      tdcError_ROSSummary = 13;
```

Comment:
please define enums for these magic numbers here and below to help with consistency/readability

Reply:
@davidlange6 We don’t see that enums would improve readability significantly as the code in the range [1] is rather compact and is well readable … [1] However we can improve this in a next PR addressing only DQM modules, for the moment we are only trying to make the least impact on the DQM clients and tests due to the new unpacker

https://github.com/jfernan2/cmssw/blob/c0d0800ff6acfaad3919cbb172c05e33e8451023/DQM/DTMonitorModule/src/DTDataIntegrityTask.cc#L858-L932

Reply:
hi @jfernan2 - its true if you have some idea what 13 (as an example) means and thus associate it elsewhere in the code. Likely the next person to modify this code won't. Some documentation is needed - easiest solution is typically enums as they avoid having to document things in multiple places and can be self description.

Reply:
Hi @davidlange6 the next person editing this piece code will be me, however I don't understand how this would ease the code. Those numbers are just bin positions to be filled in a histogram after the check is perfomed. The description is in each switch decission, and they could likely change in next iteration if other bins are introduced